### ISSUE TITLE
getModules never memoizes properly, causing performance lag/issues

### ISSUE THREAD

**Issue by tofumatt** [google/site-kit-wp#2691]
## Bug Description

Currently the `getModules` function memoises a function with an object that will never be the same: https://github.com/google/site-kit-wp/blob/f1ff73b9b078d7060af02c661f4a15ba76dd160f/assets/js/googlesitekit/modules/datastore/modules.js#L491-L495

@aaemnnosttv and I looked into this and it causes massive performance problems on the dashboard, especially when changing the date range—even when the requests are cached.

The fix is straightforward: supply `normalizeModules` with `serverDefinitions` and `clientDefinitions` rather than the `merge()`'d object, which will never be the same object for memoisation.

## Steps to reproduce

1. Go to the Dashboard in development mode
2. Click on the date range changer and select any date range other than the current one
3. Notice a several-second lag

![image](https://user-images.githubusercontent.com/90871/105886056-dfbef680-6001-11eb-8ec4-b9e2cc0e37d2.png)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* `getModules()` should not take several seconds to run when changing between date ranges
* The changing of the date range on the dashboard should be notably faster (by 1s+) compared to `develop`

## Implementation Brief

* Supply `normalizeModules` with `serverDefinitions` and `clientDefinitions` rather than the `merge()`'d object, which will never be the same object for memoisation.
* Merge #2690.

### Test Coverage

* No test coverage needed.

### Visual Regression Changes

* No VRT changes for this issue.

## QA Brief

1. Go to the Dashboard in development mode
2. Click on the date range changer and select any date range other than the current one
3. There should be no noticeable lag.

Check, in the Firefox or Chrome profiler, for `getModules()` calls. They should not take up a double-digit percentage of time when changing between date ranges.

## Changelog entry

* Fix performance lag across Site Kit screens due to a problem in the `getModules` selector of the `core/modules` store.

**Comment by felixarntz** [comment]
IB ✅ 

Awesome find!

**Comment by asvinb** [comment]
@tofumatt I did see a noticeable lag when changing to a new date range, most likely because of data being fetched in the background (using Chrome). However once the data is cached, there is no lag. I also checked in Devtools and there is no difference:
![image](https://user-images.githubusercontent.com/31818/106129873-1cadf900-617a-11eb-9864-b7b4c45a1dab.png)

If you confirm the lag is ok when fetching data, then i'll approve the ticket.
